### PR TITLE
gh-111926: Make weakrefs thread-safe in free-threaded builds

### PR DESCRIPTION
**Pull request by mpage**
This makes weakrefs thread-safe in free-threaded builds. Please see the comment at the beginning of `Objects/weakrefobject.c` for an explanation of the approach.

Note that this only affects free-threaded builds. Apart from some minor refactoring, the added code is all either gated by `ifdef`s or is a no-op (e.g. `Py_BEGIN_CRITICAL_SECTION`).

* Issue: gh-111926

### CODE REVIEW

File: Include/internal/pycore_interp.h [hunk 1]
```diff
@@ -59,6 +59,12 @@ struct _stoptheworld_state {
     PyThreadState *requester; // Thread that requested the pause (may be NULL).
 };
 
+#ifdef Py_GIL_DISABLED
+// This should be prime but otherwise the choice is arbitrary. A larger value
+// increases concurrency at the expense of memory.
+#define NUM_WEAKREF_LIST_LOCKS 127
+#endif
```
Comment (colesbury) [inline Include/internal/pycore_interp.h]: The style I've seen recommended in my PRs is to indent the `#define` here:

```suggestion
#ifdef Py_GIL_DISABLED
// This should be prime but otherwise the choice is arbitrary. A larger value
// increases concurrency at the expense of memory.
#  define NUM_WEAKREF_LIST_LOCKS 127
#endif
```

File: Include/internal/pycore_weakref.h [hunk 2]
```diff
@@ -30,53 +48,82 @@ static inline int _is_dead(PyObject *obj)
...
     }
-    Py_END_CRITICAL_SECTION();
     return ret;
 }
 
+// NB: In free-threaded builds the weakref list lock for the referenced object
+// must be held around calls to this function.
 extern Py_ssize_t _PyWeakref_GetWeakrefCount(PyWeakReference *head);
 
+extern Py_ssize_t _PyWeakref_GetWeakrefCountThreadsafe(PyObject *obj);
```
Comment (colesbury) [inline Include/internal/pycore_weakref.h]: Can we combine these two functions? Just have `_PyWeakref_GetWeakrefCount(PyObject *obj)` and make it thread-safe.

File: Python/pystate.c [hunk 3]
```diff
@@ -617,6 +625,9 @@ init_interpreter(PyInterpreterState *interp,
     _PyType_InitCache(interp);
 #ifdef Py_GIL_DISABLED
     _Py_brc_init_state(interp);
+    for (int i = 0; i < NUM_WEAKREF_LIST_LOCKS; i++) {
+        interp->weakref_locks[i] = (PyMutex){0};
+    }
```
Comment (colesbury) [inline Python/pystate.c]: We don't need to explicitly initialize these. `PyInterpreterState` is zero initialized and we don't generally explicitly initialize fields to zero.

File: Objects/weakrefobject.c [hunk 4]
```diff
@@ -48,6 +128,59 @@ new_weakref(PyObject *ob, PyObject *callback)
...
+clear_weakref_lock_held(PyWeakReference *self, PyObject **callback)
+{
+    // TODO: Assert locks are held or world is stopped
+    if (_Py_atomic_load_ptr(&self->wr_object) != Py_None) {
+        PyWeakReference **list = GET_WEAKREFS_LISTPTR(self->wr_object);
+        if (*list == self)
+            /* If 'self' is the end of the list (and thus self->wr_next ==
+               NULL) then the weakref list itself (and thus the value of *list)
+               will end up being set to NULL. */
+            _Py_atomic_store_ptr(list, self->wr_next);
```
Comment (colesbury) [inline Objects/weakrefobject.c]: Why does this use an atomic operation? Isn't `wr_next` protected by the striped locks?

File: Objects/weakrefobject.c [hunk 5]
```diff
@@ -48,6 +128,59 @@ new_weakref(PyObject *ob, PyObject *callback)
...
 }
 
+#ifdef Py_GIL_DISABLED
+
+// Clear the weakref and steal its callback into `callback`, if provided.
+static void
+clear_weakref_lock_held(PyWeakReference *self, PyObject **callback)
+{
+    // TODO: Assert locks are held or world is stopped
+    if (_Py_atomic_load_ptr(&self->wr_object) != Py_None) {
```
Comment (colesbury) [inline Objects/weakrefobject.c]: Does this need to be atomic? Don't we only modify it with the lock held?

File: Objects/weakrefobject.c [hunk 6]
```diff
@@ -1,10 +1,73 @@
...
+#include "pycore_lock.h"
 #include "pycore_modsupport.h"    // _PyArg_NoKwnames()
 #include "pycore_object.h"        // _PyObject_GET_WEAKREFS_LISTPTR()
 #include "pycore_pyerrors.h"      // _PyErr_ChainExceptions1()
+#include "pycore_pystate.h"
 #include "pycore_weakref.h"       // _PyWeakref_GET_REF()
 
+#ifdef Py_GIL_DISABLED
+/*
+ * Thread-safety for free-threaded builds
```
Comment (colesbury) [inline Objects/weakrefobject.c:12]: I think there is more background and detail here than is helpful. In the spirit of "[economy of expression](https://devguide.python.org/documentation/style-guide/#economy-of-expression)" let's trim this to the essential point:  the weakref linked list and `wr_object` are protected by a striped lock keyed by `wr_object`'s address.

File: Objects/weakrefobject.c [hunk 7]
```diff
@@ -48,6 +128,59 @@ new_weakref(PyObject *ob, PyObject *callback)
...
+    }
+    if (callback != NULL) {
+        *callback = self->wr_callback;
+        self->wr_callback = NULL;
+    }
+}
+
+// Clear the weakref while the world is stopped
+static void
+gc_clear_weakref(PyWeakReference *self)
```
Comment (colesbury) [inline Objects/weakrefobject.c]: Can these functions (`clear_weakref_lock_held`, `gc_clear_weakref`, `clear_weakref`) be combined with their default build implementations?

File: Objects/weakrefobject.c [hunk 8]
```diff
@@ -276,64 +426,157 @@ insert_head(PyWeakReference *newref, PyWeakReference **list)
...
+                     Py_TYPE(obj)->tp_name);
+        return NULL;
+    }
+    if (callback == Py_None)
+        callback = NULL;
+
+    PyWeakReference **list = GET_WEAKREFS_LISTPTR(obj);
+    PyWeakReference *ref, *proxy;
+
+    LOCK_WEAKREFS(obj);
```
Comment (colesbury) [inline Objects/weakrefobject.c]: This code structure seems more complicated than necessary because we lock twice. We only re-use weakreferences if they have no callback and one of three concrete types. Those three concrete types should be safe to allocate while holding the lock, even though it's not safe in general to allocate arbitrary types while holding the lock.

Let's restructure this so that we don't need the double checking. Something like:

1) First, check the type and callback to determine if we can reuse a weakref instance
2) If so, the with the lock held try to reuse an instance or allocate and insert a new one. Importantly the allocation and insertion happens without releasing the lock so we don't need to double check it.
3) Otherwise, allocate and return a new instance without acquiring the lock

File: Objects/weakrefobject.c [hunk 9]
```diff
@@ -276,64 +426,157 @@ insert_head(PyWeakReference *newref, PyWeakReference **list)
...
+        return NULL;
+    }
+
+    PyWeakReference *cand = NULL;
+    if (type == &_PyWeakref_RefType) {
+        cand = ref;
+    }
+    if (type == &_PyWeakref_ProxyType) {
+        cand = proxy;
+    }
```
Comment (colesbury) [inline Objects/weakrefobject.c]: Doesn't this also need to handle `_PyWeakref_CallableProxyType`?

File: Include/object.h [hunk 10]
```diff
@@ -766,6 +766,7 @@ PyAPI_FUNC(void) _Py_NegativeRefcount(const char *filename, int lineno,
                                       PyObject *op);
 PyAPI_FUNC(void) _Py_INCREF_IncRefTotal(void);
 PyAPI_FUNC(void) _Py_DECREF_DecRefTotal(void);
+PyAPI_FUNC(void) _Py_IncRefTotal(PyInterpreterState *);
```
Comment (colesbury) [inline Include/object.h]: I don't think we want to move this

File: Modules/_weakref.c [hunk 11]
```diff
@@ -86,26 +82,57 @@ Return a list of all weak reference objects pointing to 'object'.
...
     PyWeakReference **list = GET_WEAKREFS_LISTPTR(object);
-    Py_ssize_t count = _PyWeakref_GetWeakrefCount(*list);
+    Py_ssize_t count = _PyWeakref_GetWeakrefCountThreadsafe(object);
 
     PyObject *result = PyList_New(count);
     if (result == NULL) {
         return NULL;
     }
 
+#ifdef Py_GIL_DISABLED
```
Comment (colesbury) [inline Modules/_weakref.c]: This seems more complicated than necessary. Don't bother with the count, just call `PyList_Append` within the loop with the lock held.

Both the default build and the free-threaded build need to handle weakrefs with zero refcount that are not yet removed from the list. Let's add a variant of `_Py_TryIncref` that works with both builds and use it here. Something like:

```
_Py_TryIncref(PyObject *op)
{
#ifdef Py_GIL_DISABLED
 ...
#else
if (Py_REFNT(op) > 0) {
    Py_INCREF(op);
    return 1;
}
return 0;
}
```

In other words, let's have one loop for both builds and avoid the list trimming logic.

File: Include/internal/pycore_weakref.h (hunk 2 above)
Comment (mpage) [reply Include/internal/pycore_weakref.h]: That would certainly make the most sense! I'm not really sure what's acceptable with respect to changing the function signature for existing functions. I guess that really only matters for exported functions?

File: Objects/weakrefobject.c (hunk 4 above)
Comment (mpage) [reply Objects/weakrefobject.c]: We read `wr_next` without holding the lock at the beginning of `PyObject_ClearWeakRefs` to provide a fast path for the common case where an object doesn't have any weakrefs. Since `PyObject_ClearWeakRefs` is called as part of the deallocation path for any object that *could* have weakrefs I think it makes sense to have a fast path.

File: Objects/weakrefobject.c (hunk 5 above)
Comment (mpage) [reply Objects/weakrefobject.c]: > Don't we only modify it with the lock held?

We do, but we read it without holding the lock (see my reply above).

File: Objects/weakrefobject.c (hunk 4 above)
Comment (colesbury) [reply Objects/weakrefobject.c]: Ok, that makes sense to me.

File: Objects/weakrefobject.c (hunk 8 above)
Comment (mpage) [reply Objects/weakrefobject.c]: > Those three concrete types should be safe to allocate while holding the lock

I'm not sure that's true. `type` can be a subtype of `weakref`. Is it possible for subtypes to override `tp_alloc` with something that wouldn't be safe to call with the lock held?

Comment (colesbury) [reply Objects/weakrefobject.c]: We don't reuse instances if `type` is a subtype of `weakref`. We only consider reuse if the type is *exactly* `_PyWeakref_RefType`, `_PyWeakref_ProxyType`, or `_PyWeakref_CallableProxyType`:

https://github.com/python/cpython/blob/29829b58a8328a7c2ccacaa74c1d7d120a5e5ca5/Objects/weakrefobject.c#L240-L246

If `type` is a subtype (i.e., not one of those three exact types), then we should allocate first (without the lock) and then insert unconditionally (with the lock).

File: Objects/weakrefobject.c (hunk 5 above)
Comment (colesbury) [reply Objects/weakrefobject.c]: The reads that don't hold the lock should continue to use `_Py_atomic_load_ptr`, but the functions that hold the lock should use plain, non-atomic reads. We'll get better, more useful TSAN coverage if we don't use atomic loads unnecessarily.

File: Include/object.h (hunk 10 above)
Comment (mpage) [reply Include/object.h]: This is necessary because `_Py_TryIncref` ends up using it in debug builds and `_Py_TryIncref` gets inlined into `_PyWeakref_GET_REF`, which in turn gets inlined into the sqlite and ssl modules.

An alternative would be to update those modules to use `PyWeakref_GetRef`. I think that might also let us get rid of the export for `_PyMutex_LockTimed`.

File: Objects/weakrefobject.c (hunk 8 above)
Comment (mpage) [reply Objects/weakrefobject.c]: Oh yeah, of course. Sorry I failed to fully comprehend your original comment.

File: Modules/_weakref.c (hunk 11 above)
Comment (mpage) [reply Modules/_weakref.c]: I *think* that's safe... my concern is that we aren't using a critical section to lock the weakref lock, so we potentially risk deadlock by doing so.

File: Objects/weakrefobject.c (hunk 9 above)
Comment (mpage) [reply Objects/weakrefobject.c]: Yep, good catch!

**Comment by mpage** [comment]
@colesbury - Would you take another look at this when you get a chance, please?

File: Modules/_sqlite/blob.c [hunk 12]
```diff
@@ -102,8 +101,8 @@ pysqlite_close_all_blobs(pysqlite_Connection *self)
 {
     for (int i = 0; i < PyList_GET_SIZE(self->blobs); i++) {
         PyObject *weakref = PyList_GET_ITEM(self->blobs, i);
-        PyObject *blob = _PyWeakref_GET_REF(weakref);
-        if (blob == NULL) {
+        PyObject *blob;
+        if (!PyWeakref_GetRef(weakref, &blob)) {
```
Comment (colesbury) [inline Modules/_sqlite/blob.c:105]: Are these changes necessary? If not, let's keep them as is.

File: Modules/_sqlite/connection.c [hunk 13]
```diff
@@ -1065,9 +1064,11 @@ static void _pysqlite_drop_unused_cursor_references(pysqlite_Connection* self)
 
     for (Py_ssize_t i = 0; i < PyList_Size(self->cursors); i++) {
         PyObject* weakref = PyList_GetItem(self->cursors, i);
-        if (_PyWeakref_IS_DEAD(weakref)) {
+        PyObject* obj;
+        if (!PyWeakref_GetRef(weakref, &obj)) {
             continue;
         }
+        Py_DECREF(obj);
```
Comment (colesbury) [inline Modules/_sqlite/connection.c]: I don't love this rewrite to use `PyWeakref_GetRef` because now you have introduced a `Py_DECREF()` that possibly can call arbitrary code via a destructor.

File: Modules/_weakref.c [hunk 14]
```diff
@@ -86,26 +79,39 @@ Return a list of all weak reference objects pointing to 'object'.
...
+            else {
+                // Undo our _Py_TryIncref. This is safe to do with the lock
+                // held in free-threaded builds; the list holds a reference to
+                // curobj so we're guaranteed not to invoke the destructor.
+                Py_DECREF(curobj);
+            }
+        }
         current = current->wr_next;
     }
+    UNLOCK_WEAKREFS(object);
```
Comment (colesbury) [inline Modules/_weakref.c:114]: Nice. I think this fixes some bugs with `_weakref.getweakrefs` returning dead references.

File: Objects/weakrefobject.c [hunk 15]
```diff
@@ -49,25 +88,19 @@ new_weakref(PyObject *ob, PyObject *callback)
...
-clear_weakref(PyWeakReference *self)
+clear_weakref_lock_held(PyWeakReference *self, PyObject **callback)
 {
-    PyObject *callback = self->wr_callback;
-
+    // TODO: Assert locks are held or world is stopped
     if (self->wr_object != Py_None) {
         PyWeakReference **list = GET_WEAKREFS_LISTPTR(self->wr_object);
-
         if (*list == self)
```
Comment (colesbury) [inline Objects/weakrefobject.c]: There are enough changes to this function that I think adding the missing `{` `}` to if statements makes sense.

File: Objects/weakrefobject.c [hunk 16]
```diff
@@ -942,68 +1022,73 @@ PyObject_ClearWeakRefs(PyObject *object)
...
+    }
+
+    PyObject *exc = PyErr_GetRaisedException();
+    PyObject *tuple = PyTuple_New(num_weakrefs * 2);
+    if (tuple == NULL) {
+        _PyErr_ChainExceptions1(exc);
+        return;
+    }
+
+    int num_items = 0;
```
Comment (colesbury) [inline Objects/weakrefobject.c]: Let's use `Py_ssize_t` for sizes instead of `int`

File: Objects/weakrefobject.c [hunk 17]
```diff
@@ -942,68 +1022,73 @@ PyObject_ClearWeakRefs(PyObject *object)
...
         }
-        assert(!PyErr_Occurred());
-        PyErr_SetRaisedException(exc);
+        done = (*list == NULL);
+        UNLOCK_WEAKREFS(object);
+
+        Py_XDECREF(callback);
     }
+
+    for (int i = 0; i < num_items; i += 2) {
```
Comment (colesbury) [inline Objects/weakrefobject.c]: Same here: `Py_ssize_t`

File: Objects/weakrefobject.c [hunk 18]
```diff
@@ -942,68 +1022,73 @@ PyObject_ClearWeakRefs(PyObject *object)
...
-            Py_ssize_t i = 0;
-
-            tuple = PyTuple_New(count * 2);
-            if (tuple == NULL) {
-                _PyErr_ChainExceptions1(exc);
-                return;
-            }
+        done = (*list == NULL) || !is_basic_ref_or_proxy(*list);
+        UNLOCK_WEAKREFS(object);
+        Py_XDECREF(callback);
```
Comment (colesbury) [inline Objects/weakrefobject.c]: `callback` must be `NULL` here because we only clear basic refs. Let's:

1) Move the `PyObject *callback` definition inside the if statements scope
2) Assert that it's `NULL` instead of decrefing it

File: Objects/weakrefobject.c [hunk 19]
```diff
@@ -1017,10 +1102,29 @@ _PyStaticType_ClearWeakRefs(PyInterpreterState *interp, PyTypeObject *type)
...
+    for (int done = 0; !done;) {
+        PyObject *callback = NULL;
+        LOCK_WEAKREFS(type);
+        if (*list != NULL) {
+            clear_weakref_lock_held((PyWeakReference*)*list, &callback);
+        }
+        done = (*list == NULL);
+        UNLOCK_WEAKREFS(type);
+        Py_XDECREF(callback);
+    }
```
Comment (colesbury) [inline Objects/weakrefobject.c]: This is called at shutdown when there is only one thread left. We don't need to do any locking. I think the following code should be sufficient:

```
    PyObject **list = _PyStaticType_GET_WEAKREFS_LISTPTR(state);
    while (*list) {
        _PyWeakref_ClearRef((PyWeakReference *)*list);
    }
```

File: Objects/weakrefobject.c [hunk 20]
```diff
@@ -776,104 +920,40 @@ _PyWeakref_CallableProxyType = {
...
-    if (callback == Py_None)
-        callback = NULL;
-    if (callback == NULL)
-        /* return existing weak reference if it exists */
-        result = ref;
-    if (result != NULL)
-        Py_INCREF(result);
+    if (PyCallable_Check(obj)) {
+        Py_SET_TYPE(result, &_PyWeakref_CallableProxyType);
+    }
```
Comment (colesbury) [inline Objects/weakrefobject.c]: This is icky. Can we refactor this so that:

1) The type determination is pulled up into `PyWeakref_NewProxy`
2) `new_weakref` takes the type so that it doesn't have to be changed after creation

File: Objects/weakrefobject.c [hunk 21]
```diff
@@ -276,64 +330,154 @@ insert_head(PyWeakReference *newref, PyWeakReference **list)
...
+                     Py_TYPE(obj)->tp_name);
+        return NULL;
+    }
+    if (callback == Py_None)
+        callback = NULL;
+
+    PyWeakReference **list = GET_WEAKREFS_LISTPTR(obj);
+    if ((type == &_PyWeakref_RefType) ||
+        (type == &_PyWeakref_ProxyType) ||
+        (type == &_PyWeakref_CallableProxyType)) {
```
Comment (colesbury) [inline Objects/weakrefobject.c]: Put the `{` on it's own line here (see [PEP 7](https://peps.python.org/pep-0007/#:~:text=When%20you%20break%20a%20long%20expression%20at%20a%20binary%20operator%2C%20the%20operator%20goes%20at%20the%20end%20of%20the%20previous%20line%2C%20and%20braces%20should%20be%20formatted%20as%20shown.%20E.g.%3A))

File: Objects/weakrefobject.c [hunk 22]
```diff
@@ -49,25 +88,19 @@ new_weakref(PyObject *ob, PyObject *callback)
...
- * object; it is also responsible for clearing the callback slot.
- */
+// Clear the weakref and steal its callback into `callback`, if provided.
 static void
-clear_weakref(PyWeakReference *self)
+clear_weakref_lock_held(PyWeakReference *self, PyObject **callback)
 {
-    PyObject *callback = self->wr_callback;
-
+    // TODO: Assert locks are held or world is stopped
```
Comment (colesbury) [inline Objects/weakrefobject.c]: Let's either add the assertion or remove the TODO

File: Objects/weakrefobject.c [hunk 23]
```diff
@@ -276,64 +330,154 @@ insert_head(PyWeakReference *newref, PyWeakReference **list)
...
+    else {
+        insert_after(newref, prev);
+    }
+}
+
+typedef PyWeakReference *(*weakref_alloc_fn)(PyTypeObject *, PyObject *,
+                                             PyObject *);
+
+static PyWeakReference *
+get_or_create_weakref(PyTypeObject *type, weakref_alloc_fn allocate,
```
Comment (colesbury) [inline Objects/weakrefobject.c]: I think the `weakref_alloc_fn allocate` complicates this and is avoidable:

1) The generic allocation via `type->tp_alloc` always works
2) If you want, you can use the `new_weakref()` call with the `if` and the `type->tp_alloc` code path in the `else` (`new_weakref()` is possibly a bit faster)

File: Include/cpython/weakrefobject.h [hunk 24]
```diff
@@ -30,6 +30,14 @@ struct _PyWeakReference {
     PyWeakReference *wr_prev;
     PyWeakReference *wr_next;
     vectorcallfunc vectorcall;
+
+#ifdef Py_GIL_DISABLED
+    /* Pointer to the lock used when clearing in free-threaded builds.
+     * Normally this can be derived from wr_object, but in some cases we need
+     * to lock after wr_object has been set to Py_None.
+     */
+    struct _PyMutex *weakrefs_lock;
```
Comment (colesbury) [inline Include/cpython/weakrefobject.h:39]: I suspect we could get rid of this with some effort, but maybe that's not worth it. In particular, we probably want this pattern if we get rid of the striped locks in the future.

**Review by colesbury — COMMENTED** [review]
This is looking good. Some comments inline.

Additionally, `proxy_dealloc` should remove the redundant `PyObject_GC_UnTrack` inside the if statement. It's unnecessary and the plain read of `self->wr_callback` is possibly a data race.

File: Modules/_sqlite/blob.c (hunk 12 above)
Comment (mpage) [reply Modules/_sqlite/blob.c:105]: The `_sqlite` and `_ssl` modules both use some combination of `_PyWeakref_GET_REF` and `_PyWeakref_IS_DEAD`. Those are both static inline functions that end up calling unexported functions (`_PyMutex_LockTimed` and `_Py_IncRefTotal`), so linking fails without any change.

Exporting the symbols is an alternative. I don't love it, but I guess it's OK. 

Another alternative would be to add something like `_PyWeakref_IsDead` that just proxies to `_PyWeakref_IS_DEAD` and update the appropriate callsites. That would solve the symbol visibility issues and avoid changing the semantics of the calling code.

File: Include/cpython/weakrefobject.h (hunk 24 above)
Comment (mpage) [reply Include/cpython/weakrefobject.h:39]: Yeah, I have an idea for how we could get rid of this but the correctness of the resulting code is subtle. This is easier to reason about and as you said, we're probably going to want something like this when we get rid of striped locks in the future.

File: Modules/_sqlite/blob.c (hunk 12 above)
Comment (colesbury) [reply Modules/_sqlite/blob.c:105]: Ok, in that case I think the `_PyWeakref_GET_REF` -> `PyWeakref_GetRef` is fine.

For `_PyWeakref_IS_DEAD`, I think either:

1) Remove the locking in `_PyWeakref_IS_DEAD`. We don't really it. The call is inherently racy: just because it's not dead when you call `_PyWeakref_IS_DEAD` doesn't mean it's still alive in the next instant.
2) Or add the exported `_PyWeakref_IsDead` function (defined in `pycore_weakref.h`) for `_sqlite/connection.c`

File: Objects/weakrefobject.c (hunk 19 above)
Comment (mpage) [reply Objects/weakrefobject.c]: This slightly changes the semantics - the original version decrefs the callback. Doing it this way postpones that decref until the weakref's destructor is invoked, which is probably fine.

**Review by colesbury — APPROVED** [review]
LGTM

File: Python/pystate.c [hunk 25]
```diff
@@ -506,6 +506,14 @@ _PyRuntimeState_ReInitThreads(_PyRuntimeState *runtime)
     for (size_t i = 0; i < Py_ARRAY_LENGTH(locks); i++) {
         _PyMutex_at_fork_reinit(locks[i]);
     }
+#ifdef Py_GIL_DISABLED
+    for (PyInterpreterState *interp = runtime->interpreters.head;
+         interp != NULL; interp = interp->next) {
```
Comment (erlend-aasland) [inline Python/pystate.c]:
```suggestion
         interp != NULL; interp = interp->next)
    {
```

**Review by erlend-aasland — APPROVED** [review]

**Comment by hugovk** [comment]
I've started seeing a test failure that bisects to this PR. Please see issue https://github.com/python/cpython/issues/118331.
